lib/vrl/stdlib/src/ip_aton.rs

```diff
@@ -0,0 +1,79 @@
+use std::net::Ipv4Addr;
+
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct IpAton;
+
+impl Function for IpAton {
+    fn identifier(&self) -> &'static str {
+        "ip_aton"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::BYTES,
+            required: true,
+        }]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "Example",
+            source: r#"ip_aton!("1.2.3.4")"#,
+            result: Ok("67305985"),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+
+        Ok(Box::new(IpAtonFn { value }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct IpAtonFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for IpAtonFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let ip: Ipv4Addr = self
+            .value
+            .resolve(ctx)?
+            .try_bytes_utf8_lossy()?
+            .parse()
+            .map_err(|err| format!("unable to parse IPv4 address: {}", err))?;
+
+        let mut octets = ip.octets();
+        octets.reverse();
```

Comment:
Careful with the reversing here. Being host-byte order means that it depends on the host for the actual byte order. Some platforms (notably ARM) can work in big-endian mode already so blindly reversing the octets will probably give an incorrect result in these cases. So you only want to reverse when you know that the host is little-endian.

This is the reason big-endian was "force-chosen" for networking interoperability :)

Reply:
Aha, yes, you are right. I neglected Rust is returning the octets in host order. Thanks for catching!

Reply:
Actually, I realised that for `aton` you are using `octets()`, which the documentation [1] says nothing about host/network order, presumably always returns octets from left-to-right so it is always correct:
> Returns the four eight-bit integers that make up this address.

However for `ntoa` you are using `From(u32)` which the documentation [2] says is host-byte order as you mentioned:
> Converts a host byte order u32 into an Ipv4Addr.

So, I think `aton()` is unaffected but `ntoa()` will be affected instead :(

[1] https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html#method.octets
[2] https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html#impl-From%3Cu32%3E

Reply:
Should be fixed by https://github.com/timberio/vector/pull/8042/commits/08264214277ccddacf4c22307c8cfd1812b4a6cc

Reply:
Yeah, I think you are right but also that the changes in 0826421 make the implementation more clear.

Reply:
Just checked your new commit. Great idea to use big/little-endian conversions!

```rust
Ok(i.to_be().into())
```
> Converts self to big endian from the target’s endianness.
> On big endian this is a no-op. On little endian the bytes are swapped.

That sounds perfect indeed !

However, for:
```rust
let ip = Ipv4Addr::from(i.to_ne_bytes());
```
> Return the memory representation of this integer as a byte array in native byte order.
> As the target platform’s native endianness is used, portable code should use to_be_bytes or to_le_bytes, as appropriate, instead.

I wonder why you didn't use the symmetrical conversion `from_be()` and `From(u32)` instead of bytes?

Reply:
Big disclaimer, I'm not a Rust programmer!
Apologies in advance if I'm talking nonsense :)

Reply:
Aha, yes, I like that symmetry. I think I was confusing myself with what the integer was representing. I pushed 20da6e52c

Reply:
If the tests pass, it looks quite good to me! Thanks again for doing this!

My colleagues will be very happy to be able to replace these current VRL chunks:
```
ip = split!(.dst, ".")
if ( length(ip) == 4 ) {
 .dst_ipn = to_int!(ip[0]) * 256 * 256 * 256 +
        to_int!(ip[1]) * 256 * 256 +
        to_int!(ip[2]) * 256 +
        to_int!(ip[3])
}
```
With this arguably more readable and succint one-liner:
```
.dst_ipn = ip_aton(.dst)
```
💯